lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -958,50 +1030,64 @@ def decode_cloudfunctions_function_name(function_name)
     end
 
     # Requires that record has a 'kubernetes' field.
-    def handle_container_metadata(record, entry)
-      fields = %w(namespace_id namespace_name pod_id pod_name container_name)
-      fields.each do |field|
-        field_to_label(record['kubernetes'], field, entry.metadata.labels,
-                       "#{CONTAINER_SERVICE}/#{field}")
+    def extract_container_metadata(record)
+      resource_labels = {}
+      common_labels = {}
+      %w(namespace_id pod_id container_name).each do |field|
+        resource_labels.merge!(
+          fields_to_labels(record['kubernetes'], field => field))
+      end
+      %w(namespace_name pod_name).each do |field|
+        common_labels.merge!(
+          fields_to_labels(record['kubernetes'], field =>
+                         "#{CONTAINER_SERVICE}/#{field}"))
       end
       # Prepend label/ to all user-defined labels' keys.
       if record['kubernetes'].key?('labels')
         record['kubernetes']['labels'].each do |key, value|
-          entry.metadata.labels["label/#{key}"] = value
+          common_labels["label/#{key}"] = value
         end
       end
       # We've explicitly consumed all the fields we care about -- don't litter
       # the log entries with the remaining fields that the kubernetes metadata
       # filter plugin includes (or an empty 'kubernetes' field).
       record.delete('kubernetes')
       record.delete('docker')
+      [resource_labels, common_labels]
     end
 
-    def field_to_label(record, field, labels, label)
-      return unless record.key?(field)
-      labels[label] = record[field].to_s
-      record.delete(field)
+    # For every original_label => new_label pair in the label_map, delete the
+    # original_label from the record if it exists, and extract the value to form
+    # a map with the new_label as the key.
+    def fields_to_labels(record, label_map)
+      return {} if label_map.nil? || !label_map.is_a?(Hash)
+      label_map.each_with_object({}) \
```

Comment:
Why not `inject({})`? It's shorter. Also, I find the order a bit more natural but it's personal preference. WDYT?